plenum/common/stashing_router.py

```diff
@@ -1,53 +1,166 @@
-from collections import defaultdict
+from abc import ABC, abstractmethod
 from functools import partial
-from typing import Callable, Any, Dict, List, Type, Optional, Tuple
+from typing import Callable, Any, Dict, Type, Optional, Iterable
+
+from sortedcontainers import SortedListWithKey
+
+from stp_core.common.log import getlogger
 
 DISCARD = -1
 PROCESS = 0
 STASH = 1
 
 
+class StashingQueue(ABC):
+    @abstractmethod
+    def push(self, item: Any) -> bool:
+        """
+        Try to add item to stash, returns True if successful, False otherwise
+        """
+        pass
+
+    @abstractmethod
+    def pop(self) -> Any:
+        """
+        Pop next item from queue, raise exception on failure
+        """
+        pass
+
+    @abstractmethod
+    def pop_all(self) -> Iterable:
+        """
+        Remove all items from queue and return them
+        """
+        pass
+
+
+class UnsortedStash(StashingQueue):
+    def __init__(self, limit: int):
+        self._limit = limit
+        self._data = []
+
+    def __len__(self):
+        return len(self._data)
+
+    def push(self, item: Any) -> bool:
+        if len(self._data) >= self._limit:
+            return False
+        self._data.append(item)
+        return True
+
+    def pop(self) -> Any:
+        return self._data.pop(0)
+
+    def pop_all(self) -> Iterable:
+        data = self._data
+        self._data = []
+        return data
+
+
+class SortedStash(StashingQueue):
+    def __init__(self, limit: int, key: Callable):
+        self._limit = limit
+        self._key = lambda v: key(v[0])
+        self._data = SortedListWithKey(key=self._key)
+
+    def __len__(self):
+        return len(self._data)
+
+    def push(self, item: Any) -> bool:
+        if len(self._data) >= self._limit:
+            return False
+        self._data.add(item)
+        return True
+
+    def pop(self) -> Any:
+        return self._data.pop(0)
+
+    def pop_all(self) -> Iterable:
+        data = self._data
+        self._data = SortedListWithKey(key=self._key)
+        return data
+
+
 class StashingRouter:
     Handler = Callable[..., Optional[int]]
 
-    def __init__(self):
+    def __init__(self, limit: int):
+        self._limit = limit
+        self._logger = getlogger()
         self._handlers = {}  # type: Dict[Type, StashingRouter.Handler]
-        self._queues = defaultdict(list)  # type: Dict[int, List[Tuple]]
+        self._queues = {}  # type: Dict[int, StashingQueue]
+
+    def set_sorted_stasher(self, code: int, key: Callable):
+        self._queues[code] = SortedStash(self._limit, key)
 
     def subscribe(self, message_type: Type, handler: Handler):
-        # TODO: Check double subscription?
+        # TODO: Discuss - we could throw custom logic error here, but is it really better than simple assert?
```

Comment:
But should we actually throw in production in this case? If yes I'll change this to LogicError

Reply:
Should we do any assertions at all? Can it be expected that we replace a handler for some reasons?
If it's OK to replace the handler for the message type, no asserts/exceptions at all. If it's not OK, and can lead to some unexpected behavior, then maybe LogicError is better, so we don't go into a strange situation on production?

Reply:
The only reason I see for replacing handler is override in tests (although there are other ways to do that in tests). For production code it might also be useful for some cases like putting some overrides, however I think this would eventually lead to hard to understand code and should be avoided. So I'll go for LogicError then.